### PR TITLE
Fixes #6924 - Additional validations for Setting

### PR DESCRIPTION
**Pull request by xprazak2**
Validations for Settings that are supposed to contain IP address and additional check for array setting type, although I am not sure about this one.

### CODE REVIEW

**Comment by domcleal** [comment]
Setting as WoC due to extensive test failures, also rubocop.

File: app/models/setting.rb [hunk 1]
```diff
@@ -24,6 +33,36 @@ def validate(record)
     end
   end
 
+  module IpValidationCommon
```
Comment (dLobatog) [inline app/models/setting.rb]: Please split this and move it to app/validators, this logic isn't exactly Setting's logic

File: app/models/setting.rb [hunk 2]
```diff
@@ -24,6 +33,36 @@ def validate(record)
...
+      record.errors.add attribute, _("must be a valid IP regexp") unless value.match(regex) && all_plausible_numbers(value)
+    end
+  end
+
+  class IpValidator < ActiveModel::EachValidator
+    include IpValidationCommon
+
+    def validate_each(record, attribute, value)
+      regex = /\A((\d{1,3}\.){1,3}\d{1,3})\z/
+      record.errors.add attribute, _("must be a valid IP") unless value.match(regex) && all_plausible_numbers(value)
```
Comment (dLobatog) [inline app/models/setting.rb]: Don't do your own IP validations, just use `:format => { :with => Resolv::AddressRegex }` or `IPAddress.valid?(ip)`

File: app/models/setting.rb [hunk 3]
```diff
@@ -9,13 +9,22 @@ class Setting < ActiveRecord::Base
   NONZERO_ATTRS = %w{ puppet_interval idle_timeout entries_per_page max_trend outofsync_interval }
   BLANK_ATTRS = %w{ trusted_puppetmaster_hosts login_delegation_logout_url authorize_login_delegation_auth_source_user_autocreate root_pass default_location default_organization websockets_ssl_key websockets_ssl_cert oauth_consumer_key oauth_consumer_secret }
   URI_ATTRS = %w{ foreman_url unattended_url }
+  URI_BLANK_ATTRS = %w{ login_delegation_logout_url } + URI_ATTRS
+  IP_ATTRS = %w{ libvirt_default_console_address }
+  IP_REGEXP_ATTRS = %w{ remote_addr }
 
   class UriValidator < ActiveModel::EachValidator
     def validate_each(record, attribute, value)
-      record.errors.add attribute, _("must be a valid URI") unless %w(http https).include? URI(value).scheme
```
Comment (dLobatog) [inline app/models/setting.rb:15]: I missed this one, but this code can be removed, and use `app/validators/url_schema_validator.rb`

**Comment by dLobatog** [comment]
@xprazak2 Please move the validators to their own classes, that'd make it easy to test them in isolation. After that you should just make sure the validations are setup, which you can do easily by testing whether the setting is included in the constants you define, and also by using `[allow_value](https://github.com/thoughtbot/shoulda-matchers/blob/master/lib/shoulda/matchers/active_model/allow_value_matcher.rb)`. Thanks!

**Comment by dLobatog** [comment]
Merged as 7339369dc896aa9eb3c0bdc1af7992dd6268acfd, thanks @xprazak2!

**Pull request closed without merging**